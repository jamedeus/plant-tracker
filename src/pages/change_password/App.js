import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/submit/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        // Redirect to overview if logged in successfully
        if (response.ok) {
            window.location.href = '/';
        // Show correct error message if login failed
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Change Password"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Old password</span>
                    </div>
                    <input
                        name="old_password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            oldPasswordIncorrect && "input-error"
                        )}
                        value={oldPassword}
                        onInput={() => setOldPasswordIncorrect(false)}
                        onChange={(e) => setOldPassword(e.target.value)}
                    />
                </label>
                {oldPasswordIncorrect && (
                    <span className="text-error">
                        Old password incorrect
                    </span>
                )}
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">New password</span>
                    </div>
                    <input
                        name="new_password1"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword1}
                        onChange={(e) => setNewPassword1(e.target.value)}
                    />
                </label>
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Confirm new password</span>
                    </div>
                    <input
                        name="new_password2"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword2}
                        onInput={() => setNewPasswordError(false)}
                        onChange={(e) => setNewPassword2(e.target.value)}
                    />
                </label>
                {newPasswordError && (
                    <span className="text-error">
                        {newPasswordError}
                    </span>
                )}

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                >
                    Change password
                </button>
            </form>
        </div>
    );
}

export default App;
