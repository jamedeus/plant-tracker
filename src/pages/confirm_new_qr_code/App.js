import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { sendPostRequest, parseDomContext } from 'src/util';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';
import PlantDetails from 'src/components/PlantDetails.js';
import GroupDetails from 'src/components/GroupDetails.js';
import { openErrorModal } from 'src/components/ErrorModal';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faXmark, faCheck } from '@fortawesome/free-solid-svg-icons';

const Layout = ({ type, thumbnailUrl, handleConfirm }) => {
    return (
        <div className="flex flex-col gap-8 text-center my-auto">
            <p className="text-lg font-bold">
                Is this the new QR code for your {type}?
            </p>
            {thumbnailUrl && (
                <div className="mx-auto p-4 bg-base-200 rounded-3xl">
                    <img
                        className="max-h-[50vh] rounded-xl object-contain"
                        src={thumbnailUrl}
                    />
                </div>
            )}

            {/* Confirm/cancel buttons */}
            <div className="flex gap-4 mx-auto mb-8">
                <button
                    className="btn btn-error btn-square text-white"
                    onClick={() => window.location.href = "/"}
                    title="Don't change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faXmark} />
                </button>
                <button
                    className="btn btn-success btn-square text-white"
                    onClick={handleConfirm}
                    title="Change QR code"
                >
                    <FontAwesomeIcon className="size-6" icon={faCheck} />
                </button>
            </div>
        </div>
    );
};

Layout.propTypes = {
    type: PropTypes.oneOf([
        'plant',
        'group'
    ]).isRequired,
    thumbnailUrl: PropTypes.string,
    handleConfirm: PropTypes.func.isRequired
};

function App() {
    // Load UUIDs and plant/group state from django template context
    const type = parseDomContext("type");
    const instance = parseDomContext("instance");
    const newUuid = parseDomContext("new_uuid");

    // Reload if user navigates to page by pressing back button (change QR code
    // request likely expired, replace with register or manage page)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                window.location.reload();
            }
        };

        // Add listener on mount, remove on unmount
        window.addEventListener('pageshow', handleBackButton);
        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    const handleConfirm = async () => {
        const response = await sendPostRequest(
            '/change_uuid',
            {
                uuid: instance.uuid,
                new_id: newUuid
            }
        );
        // Reload page if changed successfully
        if (response.ok) {
            window.location.reload();
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Plant/group details card shown when title is clicked
    const DetailsDropdown = useMemo(() => (
        <div className="details-card">
            <div className="card-body">
                {type === 'plant' ? (
                    <PlantDetails
                        species={instance.species}
                        pot_size={instance.pot_size}
                        description={instance.description}
                    />
                ) : (
                    <GroupDetails
                        location={instance.location}
                        description={instance.description}
                    />
                )}
            </div>
        </div>
    ), []);

    return (
        <div className="container flex flex-col mx-auto h-screen">
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title={instance.display_name}
                titleOptions={DetailsDropdown}
            />
            <Layout
                type={type}
                thumbnailUrl={instance.thumbnail}
                handleConfirm={handleConfirm}
            />
        </div>
    );
}

export default App;
