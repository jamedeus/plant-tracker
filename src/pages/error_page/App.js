import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';

function App({ errorMessage }) {
    return (
        <div
            className="container flex flex-col full-screen mx-auto items-center gap-16"
            data-testid="error-page"
        >
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title="Error"
                showScanButton={false}
            />

            <div className="flex flex-col text-center text-lg mt-[15vh]">
                {errorMessage}
            </div>

            <Link to="/" className='btn btn-accent' discover="none">
                Go to Overview
            </Link>
        </div>
    );
}

export default App;

App.propTypes = {
    errorMessage: PropTypes.string.isRequired
};
