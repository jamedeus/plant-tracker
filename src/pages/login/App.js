import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [showError, setShowError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post FormData to backend
        const formData = new FormData(formRef.current);
        const response = await fetch('/accounts/login/', {
            method: 'POST',
            body: formData,
            headers: {
                'Accept': 'application/json, text/plain, */*',
                "X-CSRFToken": Cookies.get('csrftoken')
            }
        });
        // Redirect if logged in successfully
        if (response.ok) {
            // Redirect to url in querystring if present
            const params = new URL(window.location.href).searchParams;
            if (params.get('next')) {
                window.location.href = params.get('next');
            // Redirect to overview if no querystring
            } else {
                window.location.href = '/';
            }
        // Show error text if login failed
        } else {
            setShowError(true);
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Login"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Username</span>
                    </div>
                    <input
                        name="username"
                        type="text"
                        autoCapitalize="off"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={username}
                        onInput={() => setShowError(false)}
                        onChange={e => setUsername(e.target.value)}
                    />
                </label>
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">Password</span>
                    </div>
                    <input
                        name="password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            showError && "input-error"
                        )}
                        value={password}
                        onInput={() => setShowError(false)}
                        onChange={e => setPassword(e.target.value)}
                    />
                </label>

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                    disabled={username.length < 1 || password.length < 1}
                >
                    Login
                </button>

                {showError && (
                    <span className="text-error text-center">
                        Invalid username or password
                    </span>
                )}
            </form>
        </div>
    );
}

export default App;
