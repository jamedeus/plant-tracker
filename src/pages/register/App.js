import React, { useState, useRef, useEffect, memo, useMemo } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { Tab } from '@headlessui/react';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';
import { sendPostRequest, parseDomContext } from 'src/util';
import GroupDetailsForm from 'src/components/GroupDetailsForm';
import PlantDetailsForm from 'src/components/PlantDetailsForm';
import { openErrorModal } from 'src/components/ErrorModal';
import { FaXmark, FaCheck } from 'react-icons/fa6';
import { DateTime } from 'luxon';

const Form = memo(function Form({ setVisibleForm, plantFormRef, groupFormRef, showTabs, defaultValues }) {
    return (
        <Tab.Group onChange={(index) => setVisibleForm(index)}>
            {showTabs &&
                <Tab.List className="tab-group">
                    <Tab className={({ selected }) => clsx(
                        'tab-option',
                        selected && 'tab-option-selected'
                    )}>
                        Plant
                    </Tab>
                    <Tab className={({ selected }) => clsx(
                        'tab-option',
                        selected && 'tab-option-selected'
                    )}>
                        Group
                    </Tab>
                </Tab.List>
            }

            <Tab.Panels className="my-4 md:my-8">
                <Tab.Panel>
                    <PlantDetailsForm
                        formRef={plantFormRef}
                        name={defaultValues.name || ""}
                        species={defaultValues.species || ""}
                        pot_size={defaultValues.pot_size || ""}
                        description={defaultValues.description || ""}
                    />
                </Tab.Panel>
                <Tab.Panel>
                    <GroupDetailsForm
                        formRef={groupFormRef}
                    />
                </Tab.Panel>
            </Tab.Panels>
        </Tab.Group>
    );
});

Form.propTypes = {
    setVisibleForm: PropTypes.func.isRequired,
    plantFormRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
    ]).isRequired,
    groupFormRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
    ]).isRequired,
    showTabs: PropTypes.bool.isRequired,
    defaultValues: PropTypes.object.isRequired
};

const ConfirmPrompt = ({
    prompt,
    photo,
    photoAltText,
    handleConfirm,
    handleReject,
    confirmButtonTitle,
    rejectButtonTitle
}) => {
    return (
        <div className="flex flex-col full-screen justify-center text-center gap-8 px-4">
            <p className="text-lg font-bold mt-auto mb-8">
                {prompt}
            </p>
            {photo && (
                <div className="mx-auto p-4 bg-base-200 rounded-3xl">
                    <img
                        className="max-h-[50vh] rounded-xl object-contain"
                        src={photo}
                        alt={photoAltText}
                        draggable={false}
                    />
                </div>
            )}

            {/* Confirm/cancel buttons */}
            <div className="flex gap-4 mx-auto mt-8 mb-auto">
                <button
                    className="btn h-12 btn-error btn-square text-white"
                    onClick={handleReject}
                    title={rejectButtonTitle}
                >
                    <FaXmark className="size-6" />
                </button>
                <button
                    className="btn h-12 btn-success btn-square text-white"
                    onClick={handleConfirm}
                    title={confirmButtonTitle}
                >
                    <FaCheck className="size-6" />
                </button>
            </div>
        </div>
    );
};

ConfirmPrompt.propTypes = {
    prompt: PropTypes.string.isRequired,
    photo: PropTypes.string,
    photoAltText: PropTypes.string,
    handleConfirm: PropTypes.func.isRequired,
    handleReject: PropTypes.func.isRequired,
    confirmButtonTitle: PropTypes.string.isRequired,
    rejectButtonTitle: PropTypes.string.isRequired
};

function App() {
    // Load context set by django template
    const newID = parseDomContext("new_id");
    const dividingFrom = parseDomContext("dividing_from");
    const changingQrCode = parseDomContext("changing_qr_code");

    // Default form values (only used when dividing existing plant)
    const defaultValues = useMemo(() => {
        if (dividingFrom) {
            const parentName = dividingFrom.plant_details.display_name;
            const today = DateTime.now().toFormat('MMMM d, yyyy');
            return {
                name: `${dividingFrom.plant_details.display_name} prop`,
                species: dividingFrom.plant_details.species,
                pot_size: dividingFrom.plant_details.pot_size,
                description: `Divided from ${parentName} on ${today}`,
            };
        } else {
            // No defaults if registering new plant with no parent
            return {};
        }
    }, []);

    // Show confirmation screen if changing QR code for existing plant
    const [showConfirmQr, setShowConfirmQr] = useState(changingQrCode ? true : false);
    // Show confirmation screen if dividing from existing plant
    const [showConfirmDivide, setShowConfirmDivide] = useState(dividingFrom ? true : false);
    // Tracks user response at confirmation screen
    const [confirmedDividing, setConfirmedDividing] = useState(false);

    // Reload if user navigates to page by pressing back button (uuid may now
    // be registered, refresh will replace with manage plant/group page if so)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                window.location.reload();
            }
        };

        // Add listener on mount, remove on unmount
        window.addEventListener('pageshow', handleBackButton);
        return () => {
            window.removeEventListener('pageshow', handleBackButton);
        };
    }, []);

    // Track visible form (changed by tabs, used to get correct endpoint)
    // Set to 0 for plant form, 1 for group form
    const [visibleForm, setVisibleForm] = useState(0);

    const plantFormRef = useRef(null);
    const groupFormRef = useRef(null);

    // Disable save button if form invalid (field exceeded length limit)
    const [formIsValid, setFormIsValid] = useState(true);
    const onInput = () => {
        const form = visibleForm === 0 ? plantFormRef : groupFormRef;
        setFormIsValid(form.current.checkValidity());
    };

    // Change form and re-enable save button (fields clear when form changes)
    const handleFormChange = (index) => {
        setVisibleForm(index);
        setFormIsValid(true);
    };

    const handleRegister = async () => {
        // Parse all fields from visible form, set correct endpoint
        let payload, endpoint;
        if (visibleForm === 0) {
            payload = Object.fromEntries(
                new FormData(plantFormRef.current)
            );
            endpoint = '/register_plant';
        } else {
            payload = Object.fromEntries(
                new FormData(groupFormRef.current)
            );
            endpoint = '/register_group';
        }

        // If dividing from existing plant and confirmed new plant was divided:
        // add database keys from context (creates database relations between
        // new plant, parent plant, and division event)
        if (dividingFrom && confirmedDividing) {
            payload.divided_from_id = dividingFrom.plant_key;
            payload.divided_from_event_id = dividingFrom.event_key;
        }

        // Add UUID, post to backend
        payload.uuid = newID;
        const response = await sendPostRequest(endpoint, payload);
        // Show error modal if registration failed
        if (!response.ok) {
            const data = await response.json();
            openErrorModal(data.error);
        // Redirect to manage page if successfully registered
        } else if (response.redirected) {
            window.location.href = response.url;
        } else {
            const responseData = await response.json();
            openErrorModal(responseData);
        }
    };

    const handleConfirmNewQrCode = async () => {
        const response = await sendPostRequest(
            '/change_uuid',
            {
                uuid: changingQrCode.instance.uuid,
                new_id: changingQrCode.new_uuid
            }
        );
        // Reload page if changed successfully
        if (response.ok) {
            window.location.reload();
        } else {
            const error = await response.json();
            openErrorModal(JSON.stringify(error));
        }
    };

    // Top left corner dropdown options
    const DropdownMenuOptions = useMemo(() => <NavbarDropdownOptions />, []);

    return (
        <div className="container flex flex-col mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title='Registration'
            />

            {(() => {
                if (showConfirmQr) {
                    return (
                        <ConfirmPrompt
                            prompt={`Is this the new QR code for your ${changingQrCode.type}?`}
                            photo={changingQrCode.preview}
                            photoAltText={`${changingQrCode.instance.display_name} photo`}
                            handleConfirm={handleConfirmNewQrCode}
                            handleReject={() => setShowConfirmQr(false)}
                            confirmButtonTitle={"Change QR code"}
                            rejectButtonTitle={"Don't change QR code"}
                        />
                    );

                } else if (showConfirmDivide) {
                    return (
                        <ConfirmPrompt
                            prompt={`Was this plant divided from ${dividingFrom.plant_details.display_name}?`}
                            photo={dividingFrom.default_photo.preview}
                            photoAltText={`${dividingFrom.plant_details.display_name} photo`}
                            handleConfirm={() => {
                                setShowConfirmDivide(false);
                                setConfirmedDividing(true);
                            }}
                            handleReject={() => {
                                setShowConfirmDivide(false);
                                setConfirmedDividing(false);
                            }}
                            confirmButtonTitle={"Plant was divided"}
                            rejectButtonTitle={"Plant was NOT divided"}
                        />
                    );

                } else {
                    return (
                        <div
                            className="flex flex-col w-96 max-w-[100vw] px-4 mb-8"
                            onInput={onInput}
                        >
                            <Form
                                setVisibleForm={handleFormChange}
                                plantFormRef={plantFormRef}
                                groupFormRef={groupFormRef}
                                showTabs={confirmedDividing ? false : true}
                                defaultValues={confirmedDividing ? defaultValues : {}}
                            />

                            <button
                                className="btn btn-accent mx-auto"
                                disabled={!formIsValid}
                                onClick={handleRegister}
                            >
                                Save
                            </button>
                        </div>
                    );
                }
            })()}
        </div>
    );
}

export default App;
