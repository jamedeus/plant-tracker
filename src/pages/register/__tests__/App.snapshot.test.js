import createMockContext from 'src/testUtils/createMockContext';
import bulkCreateMockContext from 'src/testUtils/bulkCreateMockContext';
import { PageWrapper } from 'src/index';
import App from '../App';
import { mockContext } from './mockContext';

describe('App', () => {
    it('matches snapshot', () => {
        // Create mock state objects
        bulkCreateMockContext(mockContext);
        createMockContext('user_accounts_enabled', true);

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });
});
