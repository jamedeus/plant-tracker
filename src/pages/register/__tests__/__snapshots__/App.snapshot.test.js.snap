// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="new_id"
    >
      "0640ec3b-1bed-4b15-a078-d6e7ec66be12"
    </div>
    <div
      id="species_options"
    >
      ["Parlor Palm","Spider Plant","Calathea"]
    </div>
    <div
      id="user_accounts_enabled"
    >
      true
    </div>
    <div>
      <div
        class="container flex flex-col mx-auto items-center mb-8"
      >
        <div
          class="navbar bg-base-100 mb-4 sticky top-0 z-99"
        >
          <div
            class="dropdown justify-start min-w-12"
          >
            <div
              aria-label="Navigation menu"
              class="btn btn-ghost btn-circle size-12"
              role="button"
              tabindex="0"
            >
              <svg
                class="size-5"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 6h16M4 12h16M4 18h7"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
            <ul
              class="dropdown-content menu z-90 shadow-sm bg-base-300 rounded-box mt-3"
              tabindex="0"
            >
              <li>
                <a
                  href="/"
                >
                  Overview
                </a>
              </li>
              <li>
                <a
                  href="/accounts/profile/"
                >
                  User profile
                </a>
              </li>
              <li>
                <button>
                  Light mode
                </button>
              </li>
            </ul>
          </div>
          <div
            class="mx-auto shrink min-w-0"
            title="Scroll to top"
          >
            <div
              class="dropdown dropdown-center w-full"
            >
              <a
                class="btn btn-ghost w-full px-0 text-nowrap"
                role="button"
                tabindex="0"
              >
                <span
                  class="px-4 leading-[2.75rem] truncate"
                  style="font-size: 14px;"
                >
                  Registration
                </span>
              </a>
              <div
                class="dropdown-content z-1 flex"
                tabindex="0"
              />
            </div>
          </div>
          <div
            class="justify-end min-w-12"
          />
        </div>
        <div
          class="flex flex-col w-96 max-w-[100vw] px-4"
        >
          <div
            aria-orientation="horizontal"
            class="tab-group"
            role="tablist"
          >
            <button
              aria-controls="headlessui-tabs-panel-:r2:"
              aria-selected="true"
              class="tab-option tab-option-selected"
              data-headlessui-state="selected"
              id="headlessui-tabs-tab-:r0:"
              role="tab"
              tabindex="0"
              type="button"
            >
              Plant
            </button>
            <button
              aria-controls="headlessui-tabs-panel-:r4:"
              aria-selected="false"
              class="tab-option"
              data-headlessui-state=""
              id="headlessui-tabs-tab-:r1:"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Group
            </button>
          </div>
          <div
            class="my-4 md:my-8"
          >
            <div
              aria-labelledby="headlessui-tabs-tab-:r0:"
              data-headlessui-state="selected"
              id="headlessui-tabs-panel-:r2:"
              role="tabpanel"
              tabindex="0"
            >
              <form
                class="flex flex-col gap-4"
                id="plantDetails"
              >
                <label>
                  <span>
                    Plant name
                  </span>
                  <input
                    class="input w-full"
                    name="name"
                    type="text"
                    value=""
                  />
                </label>
                <label
                  class="relative"
                >
                  <span>
                    Plant species
                  </span>
                  <input
                    aria-autocomplete="list"
                    aria-expanded="false"
                    autocomplete="off"
                    class="input w-full"
                    data-headlessui-state=""
                    id="headlessui-combobox-input-:r3:"
                    name="species"
                    role="combobox"
                    type="text"
                    value=""
                  />
                </label>
                <label>
                  <span>
                    Pot size
                  </span>
                  <input
                    class="input w-full"
                    inputmode="numeric"
                    name="pot_size"
                    type="text"
                    value=""
                  />
                </label>
                <label>
                  <span>
                    Description
                  </span>
                  <textarea
                    class="textarea w-full"
                    name="description"
                  />
                </label>
              </form>
            </div>
            <span
              aria-hidden="true"
              aria-labelledby="headlessui-tabs-tab-:r1:"
              id="headlessui-tabs-panel-:r4:"
              role="tabpanel"
              style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
              tabindex="-1"
            />
          </div>
          <button
            class="btn btn-accent mx-auto"
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container flex flex-col mx-auto items-center mb-8"
    >
      <div
        class="navbar bg-base-100 mb-4 sticky top-0 z-99"
      >
        <div
          class="dropdown justify-start min-w-12"
        >
          <div
            aria-label="Navigation menu"
            class="btn btn-ghost btn-circle size-12"
            role="button"
            tabindex="0"
          >
            <svg
              class="size-5"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h7"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
          <ul
            class="dropdown-content menu z-90 shadow-sm bg-base-300 rounded-box mt-3"
            tabindex="0"
          >
            <li>
              <a
                href="/"
              >
                Overview
              </a>
            </li>
            <li>
              <a
                href="/accounts/profile/"
              >
                User profile
              </a>
            </li>
            <li>
              <button>
                Light mode
              </button>
            </li>
          </ul>
        </div>
        <div
          class="mx-auto shrink min-w-0"
          title="Scroll to top"
        >
          <div
            class="dropdown dropdown-center w-full"
          >
            <a
              class="btn btn-ghost w-full px-0 text-nowrap"
              role="button"
              tabindex="0"
            >
              <span
                class="px-4 leading-[2.75rem] truncate"
                style="font-size: 14px;"
              >
                Registration
              </span>
            </a>
            <div
              class="dropdown-content z-1 flex"
              tabindex="0"
            />
          </div>
        </div>
        <div
          class="justify-end min-w-12"
        />
      </div>
      <div
        class="flex flex-col w-96 max-w-[100vw] px-4"
      >
        <div
          aria-orientation="horizontal"
          class="tab-group"
          role="tablist"
        >
          <button
            aria-controls="headlessui-tabs-panel-:r2:"
            aria-selected="true"
            class="tab-option tab-option-selected"
            data-headlessui-state="selected"
            id="headlessui-tabs-tab-:r0:"
            role="tab"
            tabindex="0"
            type="button"
          >
            Plant
          </button>
          <button
            aria-controls="headlessui-tabs-panel-:r4:"
            aria-selected="false"
            class="tab-option"
            data-headlessui-state=""
            id="headlessui-tabs-tab-:r1:"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Group
          </button>
        </div>
        <div
          class="my-4 md:my-8"
        >
          <div
            aria-labelledby="headlessui-tabs-tab-:r0:"
            data-headlessui-state="selected"
            id="headlessui-tabs-panel-:r2:"
            role="tabpanel"
            tabindex="0"
          >
            <form
              class="flex flex-col gap-4"
              id="plantDetails"
            >
              <label>
                <span>
                  Plant name
                </span>
                <input
                  class="input w-full"
                  name="name"
                  type="text"
                  value=""
                />
              </label>
              <label
                class="relative"
              >
                <span>
                  Plant species
                </span>
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  autocomplete="off"
                  class="input w-full"
                  data-headlessui-state=""
                  id="headlessui-combobox-input-:r3:"
                  name="species"
                  role="combobox"
                  type="text"
                  value=""
                />
              </label>
              <label>
                <span>
                  Pot size
                </span>
                <input
                  class="input w-full"
                  inputmode="numeric"
                  name="pot_size"
                  type="text"
                  value=""
                />
              </label>
              <label>
                <span>
                  Description
                </span>
                <textarea
                  class="textarea w-full"
                  name="description"
                />
              </label>
            </form>
          </div>
          <span
            aria-hidden="true"
            aria-labelledby="headlessui-tabs-tab-:r1:"
            id="headlessui-tabs-panel-:r4:"
            role="tabpanel"
            style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;"
            tabindex="-1"
          />
        </div>
        <button
          class="btn btn-accent mx-auto"
        >
          Save
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
